### ISSUE TITLE
A possibly faster Atom struct

### ISSUE THREAD

**Issue by diegozea** [FermiQC/Molecules.jl#4]
Hi! I have seen that you are using the following `Atom` struct:

```julia
struct Atom{I,F}
    Z::I
    mass::F
    xyz::Vector{F}
end
```

On my package, MIToS, I used the [following one](https://github.com/diegozea/MIToS.jl/blob/520213e62c00f87f198b001105bbca5b1dc5bd65/src/PDB/PDBResidues.jl):

```julia
"A `Coordinates` object is a fixed size vector with the coordinates x,y,z."
@auto_hash_equals struct Coordinates <: FieldVector{3, Float64}
    x::Float64
    y::Float64
    z::Float64
    function Coordinates(a::NTuple{3, Float64})
        new(a[1], a[2], a[3])
    end
end

"""
A `PDBAtom` object contains the information from a PDB atom, without information of the
residue. It has the following fields that you can access at any moment for query purposes:
    - `coordinates` : x,y,z coordinates, e.g. `Coordinates(109.641,73.162,42.7)`.
    - `atom` : Atom name, e.g. `"CA"`.
    - `element` : Element type of the atom, e.g. `"C"`.
    - `occupancy` : A float number with the occupancy, e.g. `1.0`.
    - `B` : B factor as a string, e.g. `"23.60"`.
"""
@auto_hash_equals struct PDBAtom
    coordinates::Coordinates
    atom::String
    element::String
    occupancy::Float64
    B::String
end
```

When I wrote it, the use of [StaticArrays](https://github.com/JuliaArrays/StaticArrays.jl) gave me a good speedup.

Maybe you can take the same `Coordiantes` struct for your `Atom` type:

```julia
struct Atom{I,F}
    Z::I
    mass::F
    xyz::Coordinates
end
```

Or free the `xyz` type as a type parameter to allow for vectors, tuples, etc...

```julia
struct Atom{I,F,C}
    Z::I
    mass::F
    xyz::C
end
```

Also, I guess that Z will always be an `Int`, which will simplify the struct:

```julia
struct Atom{M,C}
    Z::Int
    mass::M
    xyz::C
end
```

Also, there is no reason to force mass and the coordinate elements to be of the same type. Imagine that you can use [Unitful](https://painterqubits.github.io/Unitful.jl/stable/) as a package user to have meaningful, and therefore different units, for mass and space coordinates.

I hope this helps :) This would be an awesome package!

Best regards,

**Comment by gustavojra** [comment]
Hello @diegozea, thanks for your input!

I think `SVectors` can be useful, we will test the performance of it once the coordinate transformations are all coded up.

My first idea was just to select a data type for Integers `I` and floats `F` stored into `Atom`. We can, of course, make it more general so it accepts mass with units. 

As for Z, while it will always be an integer, I'd like to have the freedom to use any type of `Integer`. For example, a minimal `UInt8` is the most memory-efficient way to represent any nuclei in the periodic table, however, if one wants to fake point charges with arbitrary magnitude they might want `Int64` instead. 

Summing up, I believe the future form of `Atom` will be

```
struct Atom{I,M,F}
   Z::I
   mass::M
   xyz::SVector{F}
end
```

**Comment by diegozea** [comment]
You are welcome! That's great, `SVector{3,F}` should do the trick 😀

**Comment by gustavojra** [comment]
I'm making atoms a very generic `struct`, it should work with `SVector` now!
